arduino/smartcar/smartcar.ino

```diff
@@ -1,9 +1,264 @@
+
+        hasReachedTargetDegrees = smartcarlib::utils::getAbsolute(degreesTurnedSoFar)
+                                  >= smartcarlib::utils::getAbsolute(degrees) - 5;
+    }
+    car.setSpeed(0);
+}
+
+void go(long centimeters, float speed)
+{
+    if (centimeters == 0)
+    {
+        return;
+    }
+    // Ensure the speed is towards the correct direction
+    speed = smartcarlib::utils::getAbsolute(speed) * ((centimeters < 0) ? -1 : 1);
+    car.setSpeed(speed);
+    car.setAngle(0);
+
+    long initialDistance          = car.getDistance();
+    bool hasReachedTargetDistance = false;
+    while (!hasReachedTargetDistance)
+    {
+        car.update();
+        auto currentDistance   = car.getDistance();
+        auto travelledDistance = initialDistance > currentDistance
+                                     ? initialDistance - currentDistance
+                                     : currentDistance - initialDistance;
+        hasReachedTargetDistance = travelledDistance >= smartcarlib::utils::getAbsolute(centimeters);
+    }
+    car.setSpeed(0);
+}
+
+void ctrlHeading(char input)
+{
+    switch (input)
+    {
+    case 'a': // rotate counter-clockwise going forward
+        car.setSpeed(forwardSpeed);
+        car.setAngle(leftDegrees);
+        forward = false;
+        break;
+    case 'd': // turn clock-wise
+        car.setSpeed(forwardSpeed);
+        car.setAngle(rightDegrees);
+        forward = false;
+        break;
+    case 'w': // go ahead
+        car.setSpeed(forwardSpeed);
+        car.setAngle(0);
+        obstacleAvoidance(BrakeDistance);
+        forward = true;
+        break;
+    case 's': // go back
+        if(car.getSpeed() != 0) {
+            carBrake();
+        }
+        else {
+            car.setSpeed(backwordSpeed);
+            car.setAngle(0);
+        }   
+        forward = false;
+        break;
+    default: // if you receive something that you don't know, just stop
+        car.setSpeed(0);
+        car.setAngle(0);
+    }
+}
+
+
+void setup()
+{
+    Serial.begin(9600);
+}
+
+void loop()
+{
+    // car.disableCruiseControl();
+    if (Serial.available())
+    {
+        char input = Serial.read(); // read everything that has been received so far and log down
+                                    // the last entry
+        ctrlHeading(input);
+        if(input == 'c')
+        {
+            cruiseControl();
+        }
+    }
+
+    if(updateFlag) 
+    {
+        car.update();
+    }
+
+    if(car.getSpeed() != 0 && forward) 
+    {
+        obstacleAvoidance(BrakeDistance);
+    }
+
+    if(car.getSpeed() == 0 && cruiseFlag) 
+    {
+        if(front.getDistance() < BrakeDistance && front.getDistance() > 0)
+        {
+            rotate(degreesToTurn, speedToTurn);
+            car.setSpeed(cruiseSpeed);
+            car.setAngle(0);
+            car.update();
+        }
+    }
+
+    if((front.getDistance() > 0 && front.getDistance() < crackDistance) && car.getSpeed() < 0.01 && cruiseFlag)
```

Comment:
I am pretty sure we have talked about calling getDistance multiple times and why it's wrong.